[BUG] Runtime error when wrapper is run with Druid without Backups

**Describe the bug:**
When Druid is run with No Backups, I've observed that the wrapper process goes into Runtime error, more specifically a Segmentation fault. 

The PR #51 which was aimed at fixing data directory file permissions seems to have unintentionally caused this issue. 

<summary>See the logs</summary>
<details>
  
```
{"level":"info","ts":"2025-05-09T11:21:59.263Z","caller":"etcd-wrapper/main.go:67","msg":"Running with flags: backup-restore-ca-cert-bundle-path: , backup-restore-host-port: etcd-test-local:8080, backup-restore-tls-enabled: false, etcd-client-cert-path: , etcd-client-key-path: , etcd-ready-timeout: 0s, etcd-server-name: etcd-test-local"}
{"level":"info","ts":"2025-05-09T11:21:59.263Z","caller":"app/app.go:40","msg":"Initializing application","config":{"BackupRestore":{"HostPort":"etcd-test-local:8080","TLSEnabled":false,"CaCertBundlePath":""},"EtcdClientTLS":{"ServerName":"etcd-test-local","CertPath":"","KeyPath":""}}}
{"level":"error","ts":"2025-05-09T11:21:59.264Z","caller":"bootstrap/bootstrap.go:66","msg":"error while fetching initialization status","error":"Get \"http://etcd-test-local:8080/initialization/status\": dial tcp 127.0.0.1:8080: connect: connection refused","stacktrace":"github.com/gardener/etcd-wrapper/internal/bootstrap.(*initializer).Run\n\t/go/src/github.com/gardener/etcd-wrapper/internal/bootstrap/bootstrap.go:66\ngithub.com/gardener/etcd-wrapper/internal/app.(*Application).Setup\n\t/go/src/github.com/gardener/etcd-wrapper/internal/app/app.go:58\ngithub.com/gardener/etcd-wrapper/cmd.InitAndStartEtcd\n\t/go/src/github.com/gardener/etcd-wrapper/cmd/etcd.go:66\nmain.main\n\t/go/src/github.com/gardener/etcd-wrapper/main.go:47\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:283"}
{"level":"info","ts":"2025-05-09T11:21:59.264Z","caller":"bootstrap/bootstrap.go:68","msg":"Fetched initialization status","Status":"Unknown"}
{"level":"info","ts":"2025-05-09T11:22:00.267Z","caller":"bootstrap/bootstrap.go:68","msg":"Fetched initialization status","Status":"New"}
{"level":"error","ts":"2025-05-09T11:22:00.267Z","caller":"bootstrap/bootstrap.go:150","msg":"error in checking if exitCodeFile exists, assuming full-validation to be done.","exitCodeFilePath":"/var/etcd/data/exit_code","error":"stat /var/etcd/data/exit_code: no such file or directory","stacktrace":"github.com/gardener/etcd-wrapper/internal/bootstrap.determineValidationMode\n\t/go/src/github.com/gardener/etcd-wrapper/internal/bootstrap/bootstrap.go:150\ngithub.com/gardener/etcd-wrapper/internal/bootstrap.(*initializer).Run\n\t/go/src/github.com/gardener/etcd-wrapper/internal/bootstrap/bootstrap.go:70\ngithub.com/gardener/etcd-wrapper/internal/app.(*Application).Setup\n\t/go/src/github.com/gardener/etcd-wrapper/internal/app/app.go:58\ngithub.com/gardener/etcd-wrapper/cmd.InitAndStartEtcd\n\t/go/src/github.com/gardener/etcd-wrapper/cmd/etcd.go:66\nmain.main\n\t/go/src/github.com/gardener/etcd-wrapper/main.go:47\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:283"}
{"level":"info","ts":"2025-05-09T11:22:00.267Z","caller":"bootstrap/bootstrap.go:71","msg":"Fetched initialization status is `New`. Triggering etcd initialization with validation mode","mode":"full"}
{"level":"info","ts":"2025-05-09T11:22:01.269Z","caller":"bootstrap/bootstrap.go:68","msg":"Fetched initialization status","Status":"Successful"}
{"level":"info","ts":"2025-05-09T11:22:02.270Z","caller":"bootstrap/bootstrap.go:82","msg":"Etcd initialization succeeded"}
{"level":"info","ts":"2025-05-09T11:22:09.320Z","caller":"bootstrap/bootstrap.go:126","msg":"Fetched and written etcd configuration","path":"/home/nonroot/etcd.conf.yaml"}
[WARNING] Deprecated '--logger=capnslog' flag is set; use '--logger=zap' flag instead
2025-05-09 11:22:09.321834 W | embed: Running http and grpc server on single port. This is not recommended for production.
{"level":"info","ts":"2025-05-09T11:22:09.323Z","caller":"app/readycheck.go:116","msg":"Starting HTTP server at addr","Port No: ":9095}
{"level":"warn","ts":"2025-05-09T11:22:09.324Z","caller":"clientv3/retry_interceptor.go:63","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0x4000141680/etcd-test-local:2379","attempt":0,"error":"rpc error: code = Canceled desc = grpc: the client connection is closing"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2d84f8]

goroutine 1 [running]:
net/http.(*Server).Close(0x147e970?)
    /usr/local/go/src/net/http/server.go:3091 +0x28
github.com/gardener/etcd-wrapper/internal/app.(*Application).stopHTTPServer(...)
    /go/src/github.com/gardener/etcd-wrapper/internal/app/readycheck.go:139
github.com/gardener/etcd-wrapper/internal/app.(*Application).Start.func1()
    /go/src/github.com/gardener/etcd-wrapper/internal/app/app.go:86 +0x2c
github.com/gardener/etcd-wrapper/internal/app.(*Application).Start(0x40001d0000)
    /go/src/github.com/gardener/etcd-wrapper/internal/app/app.go:96 +0x1bc
github.com/gardener/etcd-wrapper/cmd.InitAndStartEtcd({0xd33418?, 0x400051b2c0?}, 0xc055d8?, 0xbb9498?)
    /go/src/github.com/gardener/etcd-wrapper/cmd/etcd.go:69 +0xb4
main.main()
    /go/src/github.com/gardener/etcd-wrapper/main.go:47 +0x2a4
stream closed EOF for default/etcd-test-0 (etcd)
```
</details>

**Expected behavior:**
It should work alike with Backups Enabled and Backups disabled. The runtime error should not happen.

**How To Reproduce (as minimally and precisely as possible):**

Please try running Druid with the below mentioned component versions and observe the `etcd` container logs.


**Environment (please complete the following information):**
 - Etcd wrapper version/commit ID : [used the PR commit](https://github.com/gardener/etcd-wrapper/commit/653fb7b2ab67b27446fe0b3d92e03e617f951ab5)
 - Etcd-druid version/commit ID : [Commit details](https://github.com/gardener/etcd-druid/commit/dfb0fbd555de0216b590c649c0eea94d18b2d2b9)
- Etcdbr version: [Commit details](https://github.com/gardener/etcd-backup-restore/commit/3e2d5518963195105e19490a7486cbed6db1cc65)
 - Cloud Provider [All/AWS/GCS/ABS/Swift/OSS]: NA

**Anything else we need to know?**:

--- a/internal/app/app.go
+++ b/internal/app/app.go
@@ -69,6 +69,12 @@ func (a *Application) Setup() error {
 func (a *Application) Start() error {
 	var err error
 
+	// Change file permissions for files previously created without umask 0077
+	// TODO (shreyas-s-rao): remove this temporary code in etcd-wrapper v0.8.0
+	if err = bootstrap.ChangeFilePermissions(a.cfg.Dir, 0640); err != nil {
+		return fmt.Errorf("failed to change file permissions: %w", err)
+	}
+
 	// Create etcd client for readiness probe
 	cli, err := a.createEtcdClient()
 	if err != nil {
@@ -89,12 +95,6 @@ func (a *Application) Start() error {
 			)
 		}
 	}()
-
-	// Change file permissions for files previously created without umask 0077
-	// TODO (shreyas-s-rao): remove this temporary code in etcd-wrapper v0.8.0
-	if err = bootstrap.ChangeFilePermissions(a.cfg.Dir, 0640); err != nil {
-		return fmt.Errorf("failed to change file permissions: %w", err)
-	}
 
 	// Create embedded etcd and start.
 	if err = a.startEtcd(); err != nil {

--- a/internal/bootstrap/bootstrap.go
+++ b/internal/bootstrap/bootstrap.go
@@ -7,6 +7,7 @@ package bootstrap
 import (
 	"context"
 	"errors"
+	"fmt"
 	"os"
 	"path/filepath"
 	"strings"
@@ -85,14 +86,29 @@ func (i *initializer) Run(ctx context.Context) (*embed.Config, error) {
 
 // ChangeFilePermissions changes the file permissions of all files in the given directory and its subdirectories recursively.
 func ChangeFilePermissions(dir string, mode os.FileMode) error {
+	info, err := os.Stat(dir)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return nil
+		}
+		return fmt.Errorf("error stating directory %s: %v", dir, err)
+	}
+
+	if !info.IsDir() {
+		return fmt.Errorf("path %s is not a directory", dir)
+	}
+
 	return filepath.WalkDir(dir, func(path string, d os.DirEntry, err error) error {
 		if err != nil {
-			return err
+			return fmt.Errorf("error walking the path %q: %v", path, err)
 		}
 		if d.IsDir() {
 			return nil
 		}
-		return os.Chmod(path, mode)
+		if err = os.Chmod(path, mode); err != nil {
+			return fmt.Errorf("error changing file permissions for %q: %v", path, err)
+		}
+		return nil
 	})
 }
 

--- a/internal/bootstrap/bootstrap.go
+++ b/internal/bootstrap/bootstrap.go
@@ -91,7 +91,7 @@ func ChangeFilePermissions(dir string, mode os.FileMode) error {
 		if errors.Is(err, os.ErrNotExist) {
 			return nil
 		}
-		return fmt.Errorf("error stating directory %s: %v", dir, err)
+		return fmt.Errorf("error stating directory %s: %w", dir, err)
 	}
 
 	if !info.IsDir() {
@@ -100,13 +100,13 @@ func ChangeFilePermissions(dir string, mode os.FileMode) error {
 
 	return filepath.WalkDir(dir, func(path string, d os.DirEntry, err error) error {
 		if err != nil {
-			return fmt.Errorf("error walking the path %q: %v", path, err)
+			return fmt.Errorf("error walking the path %q: %w", path, err)
 		}
 		if d.IsDir() {
 			return nil
 		}
 		if err = os.Chmod(path, mode); err != nil {
-			return fmt.Errorf("error changing file permissions for %q: %v", path, err)
+			return fmt.Errorf("error changing file permissions for %q: %w", path, err)
 		}
 		return nil
 	})

--- a/internal/bootstrap/bootstrap_test.go
+++ b/internal/bootstrap/bootstrap_test.go
@@ -29,6 +29,89 @@ func (f TestRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
 	return f(req), nil
 }
 
+func TestChangeFilePermissions(t *testing.T) {
+	table := []struct {
+		description string
+		setup       func(testDir string) string
+		mode        os.FileMode
+		expectError bool
+		verify      func(t *testing.T, testDir string, mode os.FileMode)
+	}{
+		{
+			description: "change permissions of files in a directory recursively",
+			setup: func(testDir string) string {
+				filePath := filepath.Join(testDir, "testfile")
+				err := os.WriteFile(filePath, []byte("test"), 0644)
+				if err != nil {
+					t.Fatalf("failed to create test file: %v", err)
+				}
+				subDirPath := filepath.Join(testDir, "subdir")
+				err = os.Mkdir(subDirPath, 0755)
+				if err != nil {
+					t.Fatalf("failed to create test directory: %v", err)
+				}
+				subFilePath := filepath.Join(subDirPath, "subfile")
+				err = os.WriteFile(subFilePath, []byte("test"), 0644)
+				if err != nil {
+					t.Fatalf("failed to create test file: %v", err)
+				}
+				return testDir
+			},
+			mode:        0640,
+			expectError: false,
+			verify: func(t *testing.T, testDir string, mode os.FileMode) {
+				g := NewWithT(t)
+				paths := []string{
+					filepath.Join(testDir, "testfile"),
+					filepath.Join(testDir, "subdir", "subfile"),
+				}
+				for _, path := range paths {
+					info, err := os.Stat(path)
+					g.Expect(err).To(BeNil())
+					g.Expect(info.Mode().Perm()).To(Equal(mode))
+				}
+			},
+		},
+		{
+			description: "return nil error for non-existent directory",
+			setup: func(testDir string) string {
+				return filepath.Join(testDir, "nonexistent/path")
+			},
+			mode:        0640,
+			expectError: false,
+			verify:      func(_ *testing.T, _ string, _ os.FileMode) {},
+		},
+		{
+			description: "return error when path is a file, not a directory",
+			setup: func(testDir string) string {
+				filePath := filepath.Join(testDir, "testfile")
+				err := os.WriteFile(filePath, []byte("test"), 0644)
+				if err != nil {
+					t.Fatalf("failed to create test file: %v", err)
+				}
+				return filePath
+			},
+			mode:        0640,
+			expectError: true,
+			verify:      func(_ *testing.T, _ string, _ os.FileMode) {},
+		},
+	}
+
+	for _, entry := range table {
+		t.Run(entry.description, func(t *testing.T) {
+			g := NewWithT(t)
+			testDir := createTestDir(t)
+			defer deleteTestDir(t, testDir)
+
+			path := entry.setup(testDir)
+			err := ChangeFilePermissions(path, entry.mode)
+			g.Expect(err != nil).To(Equal(entry.expectError))
+
+			entry.verify(t, testDir, entry.mode)
+		})
+	}
+}
+
 func TestCleanupExitCodeFile(t *testing.T) {
 	table := []struct {
 		description string

